File: Leetcode/python/sol279.py

# https://leetcode.com/problems/perfect-squares/

class Solution:
    def numSquares(self, n: int) -> int:
        squares = [i*i for i in range(1, int(sqrt(n) + 1))]
        memo = {}
        def helper(val):
            if val in memo:
                return memo[val]
            if val == 0:
                return 0
            if val < 0:
                return -1
            minimum = float('inf')
            for i in squares:
                res = helper(val - i)
                if res != -1:
                    minimum = min(minimum, res + 1)
            memo[val] = minimum
            return memo[val]
        return helper(n)

File: Leetcode/typescript/sol279.ts

//https://leetcode.com/problems/perfect-squares/submissions/1943906527/


function numSquares(n: number): number {
    const dp:number[] = Array(n+1).fill(Infinity);
    dp[0] = 0;
    const squares:number[] = [];
    for(let i = 1; i*i <= n; i++){
        squares.push(i*i);
    }
    for(let i = 1; i <= n; i++){
        for (const num of squares){
            let need:number = i - num;
            if(num > i){
                break;
            }
            dp[i] = Math.min(dp[i], dp[need]+1)

        }
    }
    return dp[n] == Infinity ? -1:dp[n];
};